### PR TITLE
kubeadm upgrade: Allow supplying hand migrated component configs

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Currently, kubeadm would refuse to perform an upgrade
if it detects a user supplied unsupported component config version. Hence,
users are required to manually upgrade their component configs and store them
in the config maps prior to executing `kubeadm upgrade plan` or
`kubeadm upgrade apply`.

This change introduces the ability to use the `--config` option of the
`kubeadm upgrade plan` and `kubeadm upgrade apply` commands to supply a YAML
file containing component configs to be used in place of the existing ones in
the cluster upon upgrade.

The old behavior where `--config` is used to reconfigure a cluster is still
supported. kubeadm automatically detects which behavior to use based on the
presence (or absense) of kubeadm config types (API group
`kubeadm.kubernetes.io`).

**Which issue(s) this PR fixes**:

Refs:
- kubernetes/kubeadm#1940
- kubernetes/enhancements#1381

**Special notes for your reviewer**:

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-longterm
/assign @fabriziopandini @neolit123 @ereslibre

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: Allow manually upgraded component configs to be supplied in a YAML file via the --config option during upgrade plan & apply. The old behavior of --config in which kubeadm configuration and component configs that overwrite everything cluster stored is preserved too. The behavior to use with --config is now determined based on whether kubeadm config API objects (API group "kubeadm.kubernetes.io") were supplied in the file or not.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thank you for this work @rosti 
i did a first pass and added some minor comments, but nothing outstanding.

File: cmd/kubeadm/app/componentconfigs/configset_test.go [hunk 1]
```diff
@@ -110,3 +110,139 @@ func TestFetchFromDocumentMap(t *testing.T) {
...
+				},
+			)
+
+			docmap, err := kubeadmutil.SplitYAMLDocuments([]byte(test.config))
+			if err != nil {
+				t.Fatalf("unexpected failure of SplitYAMLDocuments: %v", err)
+			}
+
+			err = FetchFromClusterWithLocalUpgrades(clusterCfg, client, docmap)
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (rosti) [reply cmd/kubeadm/app/componentconfigs/configset_test.go:225]: I am personally not very fond of this style especially when we have a more complex test validity check, than just "check if the func returned or not error as expected".

**Review by neolit123 — COMMENTED** [review]
/lgtm

**Pull request merged**